src/utils/internalLinks.js

```diff
@@ -0,0 +1,117 @@
+/* @flow strict-local */
+import type { Narrow, User, Stream } from '../types';
+import { topicNarrow, streamNarrow, groupNarrow, specialNarrow } from './narrow';
+import { isUrlOnRealm } from './url';
+import { getUserById } from '../users/userHelpers';
+import { transformToEncodedURI } from './string';
+import { NULL_USER } from '../nullObjects';
+
+export const getPathsFromUrl = (url: string = '', realm: string) => {
+  const paths = url
+    .split(realm)
+    .pop()
+    .split('#narrow/')
+    .pop()
+    .split('/');
+
+  if (paths.length > 0 && paths[paths.length - 1] === '') {
+    // url ends with /
+    paths.splice(-1, 1);
+  }
+  return paths;
+};
+
+export const isInternalLink = (url: string, realm: string): boolean =>
+  isUrlOnRealm(url, realm) ? /^(\/#narrow|#narrow)/i.test(url.split(realm).pop()) : false;
+
+export const isMessageLink = (url: string, realm: string): boolean =>
+  isInternalLink(url, realm) && url.includes('near');
+
+type LinkType = 'external' | 'home' | 'group' | 'topic' | 'stream' | 'special';
+
+export const getLinkType = (url: string, realm: string): LinkType => {
```

Comment:
Some other cases I'm curious about:
* With `/near/ID` on a stream
* A PM with one person, not a group
* A search
* A sent-by query
* A search or sent-by query, combined with a stream and/or topic, or with a group PM thread, etc.

More generally, the URLs we support in the webapp are a lot richer than this code is built to accept. A couple of the cases above should probably be fixed even with the current structure of this code, but I think this is not the right architecture and in the end we should do something that better matches how the URLs actually work.

(None of that is an obstacle to merging this PR, though! I see that it's refactoring the logic that exists, and that's a valuable thing to do. Improvements to the behavior may or may not land in this PR, but they should definitely be separate commits.)

Reply:
A PM with one person, not a group

'group' now renamed to 'pm' to reflect the correct usage.